### PR TITLE
Basic bots (medibots)

### PR DESCRIPTION
**Pull request by Ben10Omintrix**
## About The Pull Request
refactors bots into basic bots. i decided to do medibots first because they were the most complex bots and wanted to get them out of the way first. if this pr gets merged then i will be rolling out the rest of the bot refactors over the next months  

medibots can now wear any type of  hats! some of these hats can trigger unique new recorded voicelines and interactions!
chickembot
hatbot

## Why It's Good For The Game
converts medbots into basic medbots. they are now a bit more responsive than before but overall they should act similarly to how they did. also adds the basic bot ai framework which has all the generic bot behaviors that the next bots may or may not need to have in their ai. i tried my best to improve their code and turned some of their bitflags into define bitfields. this pr may need a careful review because i did it from scratch as the old bot project branch was too outdated for me to use
## Changelog
:cl:
refactor: medbots are now basic bots. please report any bugs
add: medbots can wear hats!
/:cl:

### CODE REVIEW

**Comment by nikothedude** [comment]
we love you so much ben

**Comment by SomeoneYouProbablyKnow** [comment]
What a blessed PR.
Now to make bots wearable as hats

**Comment by AdipemDragon** [comment]
oh, new voicelines are definitely going to be insteresting to hear ingame, that's for sure

**Review by MrMelbert — COMMENTED** [review]
The last attempt to move a bot to basic mob left me very unhappy to see a thousand lines of code of copy+pasted bot code, forcing us to maintain `/mob/living/simple_animal/bot` and `/mob/basic/bot` simultaneously (#70658) 

In normal basic mob cases, I wouldn't mind this - but simple bots are very commonly touched code, and having 1:1 two copies of the type creates a maintainability issue. 

I would really prefer if we just focused on the essentials of porting the bots over and did so on a project branch* so we don't have to deal with two bot types for months while we wait for them all to finish and converted

*Or alternatively, if we could arrange something with Jacquerel, San, maybe Fulp Willard to lessen the pain of this that would be nice as well

**Review by MrMelbert — COMMENTED** [review]
This PR doesn't touch `isbot` so it'll break anything that uses that

It also makes anything that iterates over `bots_list` rather fragile

File: code/datums/elements/hat_wearer.dm [hunk 1]
```diff
@@ -0,0 +1,73 @@
+/**
+ * # mobs that can wear hats!
+ */
+/datum/element/hat_wearer
+	element_flags = ELEMENT_DETACH_ON_HOST_DESTROY|ELEMENT_BESPOKE
+	argument_hash_start_idx = 2
+	///offsets of hats we will wear
+	var/list/offsets
```
Comment (MrMelbert) [inline code/datums/elements/hat_wearer.dm:8]: Not related to above but it is a great shame that cyborgs do not use this element.

File: code/__DEFINES/robots.dm [hunk 2]
```diff
@@ -300,3 +318,4 @@ DEFINE_BITFIELD(janitor_mode_flags, list(
 #define MEDIBOT_VOICED_THIS_HURTS "This hurts, my pain is real!"
 #define MEDIBOT_VOICED_THE_END "Is this the end?"
 #define MEDIBOT_VOICED_NOOO	"Nooo!"
+#define MEDIBOT_VOICED_CHICKEN "LOOK AT ME??? i am a chicken"
```
Comment (MrMelbert) [inline code/__DEFINES/robots.dm:321]:
```suggestion
#define MEDIBOT_VOICED_CHICKEN "LOOK AT ME!? I am a chicken."
```

**Comment by Ben10Omintrix** [comment]
i will do what u want

**Comment by Ben10Omintrix** [comment]
> This PR doesn't touch `isbot` so it'll break anything that uses that
> 
> It also makes anything that iterates over `bots_list` rather fragile

yes theres still alot of things i need to do before i undraft this

File: code/__DEFINES/ai/ai_blackboard.dm [hunk 3]
```diff
@@ -140,3 +140,39 @@
 #define BB_BASIC_MOB_REINFORCEMENT_TARGET "BB_basic_mob_reinforcement_target"
 /// The next time at which this mob can call for reinforcements
 #define BB_BASIC_MOB_REINFORCEMENTS_COOLDOWN "BB_basic_mob_reinforcements_cooldown"
+
+// bot keys
```
Comment (san7890) [inline code/__DEFINES/ai/ai_blackboard.dm:144]: this is gonna get ridiculous rather fast so i think it's time we make the `blackboard` folder in the `ai` folder, move the old file (the one you're currently editing) and then make a new file with the new keys in it.

File: code/__DEFINES/robots.dm [hunk 4]
```diff
@@ -104,6 +104,22 @@
...
+#define BOT_MAINTS_PANEL_OPEN (1<<0)
+///is our control panel currently open
+#define BOT_CONTROL_PANEL_OPEN (1<<1)
+
+///bitfield for our access flags
+DEFINE_BITFIELD(bot_access_flags, list(
+	"MAINTS OPEN" = BOT_MAINTS_PANEL_OPEN,
+	"CONTROL OPEN" = BOT_CONTROL_PANEL_OPEN,
+	"COVER EMAGGED" = BOT_COVER_EMAGGED,
+	"COVER HACKED" = BOT_COVER_HACKED,
```
Comment (san7890) [inline code/__DEFINES/robots.dm:120]: Shouldn't these be SCREAMING_SNAKE_CASE as well, e.g. `MAINTS_OPEN`

File: code/datums/elements/hat_wearer.dm [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+/datum/element/hat_wearer/Detach(datum/target)
+	var/obj/item/hat = (locate(/obj/item/clothing/head) in target)
+	if(hat)
+		hat.forceMove(get_turf(target))
+	UnregisterSignal(target, list(
+		COMSIG_ATOM_UPDATE_OVERLAYS,
+		COMSIG_ATOM_EXITED,
+		COMSIG_ATOM_ENTERED,
+		COMSIG_ATOM_ATTACKBY,
+		COMSIG_ATOM_HITBY,
```
Comment (san7890) [inline code/datums/elements/hat_wearer.dm:29]: We aren't registering HITBY why are we deregistering it

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 6]
```diff
@@ -0,0 +1,781 @@
...
+/mob/living/basic/bot/proc/rename(mob/user)
+	var/new_name = sanitize_name(
+		reject_bad_text(tgui_input_text(
+			user = user,
+			message = "This machine is designated [real_name]. Would you like to update its registration?",
+			title = "Name change",
+			default = real_name,
+			max_length = MAX_NAME_LEN,
+		)),
+		allow_numbers = TRUE
```
Comment (san7890) [inline code/modules/mob/living/basic/bots/_bots.dm:246]:
```suggestion
		allow_numbers = TRUE,
```

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 7]
```diff
@@ -0,0 +1,781 @@
...
+
+	if(bot_access_flags & BOT_MAINTS_PANEL_OPEN)
+		balloon_alert(user, "open the access panel!")
+		return
+
+	balloon_alert(user, "removing pAI...")
+	if(!do_after(user, 3 SECONDS, target = src) || !paicard)
+		return
+
+	user.visible_message(span_notice("[user] uses [attacking_item] to pull [paicard] out of [initial(src.name)]!"),span_notice("You pull [paicard] out of [initial(src.name)] with [attacking_item]."))
```
Comment (san7890) [inline code/modules/mob/living/basic/bots/_bots.dm:442]: multiline this while we're here?

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 8]
```diff
@@ -0,0 +1,781 @@
...
+/// Ejects the pAI remotely.
+/mob/living/basic/bot/proc/eject_pai_remote(mob/user)
+	if(!check_access(user) || !paicard)
+		return
+	speak("Ejecting personality chip.", radio_channel)
+	ejectpai(user)
+
+/mob/living/basic/bot/Login()
+	. = ..()
+	if(!. || !client)
```
Comment (san7890) [inline code/modules/mob/living/basic/bots/_bots.dm:748]: `isnull()` here instead

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 9]
```diff
@@ -0,0 +1,781 @@
...
+
+	if(door_attempt.check_access(access_card))
+		return ACCESS_ALLOWED
+	return ACCESS_DISALLOWED
+
+/mob/living/basic/bot/proc/generate_speak_list()
+	return null
+
+/mob/living/basic/bot/proc/provide_additional_access()
+	var/datum/id_trim/additional_trim = SSid_access.trim_singletons_by_path[additional_access]
```
Comment (san7890) [inline code/modules/mob/living/basic/bots/_bots.dm:779]: yea this is dogging on you in unit tests

File: code/modules/mob/living/basic/bots/medbot/medbot.dm [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+	/// How much healing do we do at a time?
+	var/heal_amount = 2.5
+	/// Start healing when they have this much damage in a category
+	var/heal_threshold = 10
+	/// What damage type does this bot support. Because the default is brute, if the medkit is brute-oriented there is a slight bonus to healing. set to "all" for it to heal any of the 4 base damage types
+	var/damage_type_healer = BRUTE
+
+	///Flags Medbots use to decide how they should be acting.
+	var/medical_mode_flags = MEDBOT_DECLARE_CRIT | MEDBOT_SPEAK_MODE
+//	Selections:  MEDBOT_DECLARE_CRIT | MEDBOT_STATIONARY_MODE | MEDBOT_SPEAK_MODE
```
Comment (san7890) [inline code/modules/mob/living/basic/bots/medbot/medbot.dm:106]:
```suggestion
	///Flags Medbots use to decide how they should be acting. Can be MEDBOT_DECLARE_CRIT | MEDBOT_STATIONARY_MODE | MEDBOT_SPEAK_MODE
	var/medical_mode_flags = MEDBOT_DECLARE_CRIT | MEDBOT_SPEAK_MODE
```

ugly unscoped comment

File: code/modules/mob/living/basic/bots/medbot/medbot.dm [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+	icon_state = "[base_icon_state][get_bot_flag(medical_mode_flags, MEDBOT_STATIONARY_MODE) ? 2 : 1]" //Bot has yellow light to indicate stationary mode.
+
+/mob/living/basic/bot/medbot/update_overlays()
+	. = ..()
+	if(skin)
+		. += "medskin_[skin]"
+
+//this is sin
+/mob/living/basic/bot/medbot/generate_speak_list()
+	return (idle_lines + wait_announcements + afterheal_announcements + near_death_announcements + emagged_announcements + tipped_announcements + untipped_announcements + worried_announcements + misc_announcements)
```
Comment (san7890) [inline code/modules/mob/living/basic/bots/medbot/medbot.dm:177]:
```suggestion
/mob/living/basic/bot/medbot/generate_speak_list()
	var/static/list/finalized_speak_list = null
	if(isnull(finalized_speak_list))
		finalized_speak_list = (idle_lines + wait_announcements + afterheal_announcements + near_death_announcements + emagged_announcements + tipped_announcements + untipped_announcements + worried_announcements + misc_announcements)
	return finalized_speak_list
```

no need to add seventy trillion lists together all da damn time

File: code/modules/mob/living/basic/bots/medbot/medbot.dm [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+	else if(damage_type_healer == HEAL_ALL_DAMAGE)
+		patient.heal_overall_damage(heal_amount)
+	else
+		patient.heal_damage_type(heal_amount = heal_amount, damagetype = damage_type_healer)
+	update_bot_mode(new_mode = BOT_IDLE)
+
+
+/mob/living/basic/bot/proc/update_bot_mode(new_mode)
+	mode = new_mode
+	update_appearance()
```
Comment (san7890) [inline code/modules/mob/living/basic/bots/medbot/medbot.dm:310]: shouldn't this be in the `_bots.dm` file

File: tools/UpdatePaths/Scripts/medbots.txt [hunk 13]
```diff
@@ -0,0 +1,6 @@
+/mob/living/simple_animal/bot/medbot : /mob/living/basic/bot/medbot{@OLD}
```
Comment (san7890) [inline tools/UpdatePaths/Scripts/medbots.txt:1]: why did you make two updatepaths files

**Review by san7890 — CHANGES_REQUESTED** [review]
you'll get to these eventually. some big oopsies though

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 14]
```diff
@@ -0,0 +1,781 @@
...
+		else
+			. += "[src]'s parts look very loose!"
+	else
+		. += "[src] is in pristine condition."
+
+	. += span_notice("Its maintenance panel is [bot_access_flags & BOT_MAINTS_PANEL_OPEN ? "open" : "closed"].")
+	. += span_info("You can use a <b>screwdriver</b> to [bot_access_flags & BOT_MAINTS_PANEL_OPEN ? "close" : "open"] it.")
+
+	if(bot_access_flags & BOT_MAINTS_PANEL_OPEN)
+		. += span_notice("Its control panel is [bot_access_flags & BOT_CONTROL_PANEL_OPEN ? "unlocked" : "locked"].")
```
Comment (JohnFulpWillard) [inline code/modules/mob/living/basic/bots/_bots.dm:332]: We only show the text that the panel is open/closed if it's open?

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 15]
```diff
@@ -0,0 +1,781 @@
...
+		var/is_sillycone = issilicon(user)
+		if(!(bot_access_flags & BOT_COVER_EMAGGED) && (is_sillycone || user.Adjacent(src)))
```
Comment (JohnFulpWillard) [inline code/modules/mob/living/basic/bots/_bots.dm:334]:
```suggestion
		if(!(bot_access_flags & BOT_COVER_EMAGGED) && (issilicon(user) || user.Adjacent(src)))
```
No point in it

File: code/modules/mob/living/basic/bots/_bots.dm [hunk 16]
```diff
@@ -0,0 +1,781 @@
...
+		if("summon")
+			bot_reset()
+			ai_controller?.set_blackboard_key(BB_BOT_SUMMON_TARGET, get_turf(user))
+			if(length(user_access))
+				access_card.set_access(user_access + prev_access) //Adds the user's access, if any.
+			mode = BOT_SUMMON
+			speak("Responding.", radio_channel)
+		if("ejectpai")
+			eject_pai_remote(user)
+	return
```
Comment (JohnFulpWillard) [inline code/modules/mob/living/basic/bots/_bots.dm:572]:
```suggestion
```
No point in it

File: code/modules/mob/living/basic/bots/medbot/medbot.dm (hunk 10 above)
Comment (JohnFulpWillard) [reply code/modules/mob/living/basic/bots/medbot/medbot.dm:106]: I think I was originally told not to do this when I added the flags

File: code/modules/mob/living/basic/bots/_bots.dm (hunk 14 above)
Comment (Ben10Omintrix) [reply code/modules/mob/living/basic/bots/_bots.dm:332]: One flag is for the maints, the other is for the control. They are not the same flag

Comment (JohnFulpWillard) [reply code/modules/mob/living/basic/bots/_bots.dm:332]: oh my god

File: code/datums/elements/hat_wearer.dm (hunk 5 above)
Comment (Ben10Omintrix) [reply code/datums/elements/hat_wearer.dm:29]: op i was planning to make it so u can throw hats on them but i forgot to implement that

File: code/modules/mob/living/basic/bots/_bots.dm (hunk 15 above)
Comment (Ben10Omintrix) [reply code/modules/mob/living/basic/bots/_bots.dm:334]: it was caching it because it uses this check again in the span info!

**Pull request merged**